File: aat-android/src/main/java/ch/bailu/aat/preferences/map/AndroidMapDirectories.java
package ch.bailu.aat.preferences.map;

import android.content.Context;

import java.io.File;
import java.util.ArrayList;

import ch.bailu.aat.preferences.Storage;
import ch.bailu.aat.util.fs.AndroidVolumes;
import ch.bailu.aat_lib.preferences.SelectionList;
import ch.bailu.aat_lib.preferences.StorageInterface;
import ch.bailu.aat_lib.preferences.map.MapDirectories;
import ch.bailu.aat_lib.preferences.map.SolidMapsForgeDirectory;
import ch.bailu.aat_lib.preferences.map.SolidMapsForgeMapFile;
import ch.bailu.aat_lib.preferences.map.SolidRenderTheme;
import ch.bailu.aat_lib.util.fs.AppDirectory;
import ch.bailu.foc.Foc;
import ch.bailu.foc.FocFactory;
import ch.bailu.foc_android.FocAndroid;
import ch.bailu.foc_android.FocAndroidFactory;

public class AndroidMapDirectories implements MapDirectories {


    private final Context context;

    public AndroidMapDirectories(Context context) {
        this.context = context;
    }

    @Override
    public ArrayList<Foc> getWellKnownMapDirs() {
        final ArrayList<Foc> dirs = new ArrayList<>(5);
        final AndroidVolumes volumes = new AndroidVolumes(context);

        for (Foc f : volumes.getVolumes()) {
            SelectionList.add_dr(dirs, f.child(SolidMapsForgeDirectory.MAPS_DIR));
            SelectionList.add_dr(dirs, f.child(AppDirectory.DIR_AAT_DATA + "/" + SolidMapsForgeDirectory.MAPS_DIR));
            SelectionList.add_dr(dirs, f.child(SolidMapsForgeDirectory.ORUX_MAPS_DIR));
        }

        return dirs;
    }

    @Override
    public Foc getDefault() {
        File external = context.getExternalFilesDir(null);
        return FocAndroid.factory(context, external.getAbsolutePath());
    }

    @Override
    public SolidMapsForgeDirectory createSolidDirectory() {
        final FocFactory foc = new FocAndroidFactory(context);
        final StorageInterface storage = new Storage(context);

        return new SolidMapsForgeDirectory(storage, foc, this);
    }

    @Override
    public SolidMapsForgeMapFile createSolidFile() {
        final FocFactory foc = new FocAndroidFactory(context);
        final StorageInterface storage = new Storage(context);

        return new SolidMapsForgeMapFile(storage, foc, this);
    }

    @Override
    public SolidRenderTheme createSolidRenderTheme() {
        return new SolidRenderTheme(createSolidDirectory(),new FocAndroidFactory(context));
    }
}

File: aat-android/src/main/kotlin/ch/bailu/aat/preferences/map/AndroidMapDirectories.kt
package ch.bailu.aat.preferences.map

import android.content.Context
import ch.bailu.aat.preferences.Storage
import ch.bailu.aat.util.fs.AndroidVolumes
import ch.bailu.aat_lib.preferences.SelectionList
import ch.bailu.aat_lib.preferences.StorageInterface
import ch.bailu.aat_lib.preferences.map.MapDirectories
import ch.bailu.aat_lib.preferences.map.SolidMapsForgeDirectoryHint
import ch.bailu.aat_lib.preferences.map.SolidMapsForgeMapFile
import ch.bailu.aat_lib.preferences.map.SolidRenderTheme
import ch.bailu.aat_lib.util.fs.AppDirectory
import ch.bailu.foc.Foc
import ch.bailu.foc.FocFactory
import ch.bailu.foc_android.FocAndroid
import ch.bailu.foc_android.FocAndroidFactory
import java.io.File

class AndroidMapDirectories(private val context: Context) : MapDirectories {
    override fun getWellKnownMapDirs(): ArrayList<Foc> {
        val result = ArrayList<Foc>()
        val volumes = AndroidVolumes(context)
        for (f in volumes.volumes) {
            SelectionList.addDr(result, f.child(MAPS_DIR))
            SelectionList.addDr(
                result,
                f.child(AppDirectory.DIR_AAT_DATA + "/" + MAPS_DIR)
            )
            SelectionList.addDr(result, f.child(ORUX_MAPS_DIR))
        }

        // app_private/files/maps (readable and on external medium)
        val files = volumes.files
        for (i in 1 until files.size) {
            SelectionList.addDr(result, files[i].child(MAPS_DIR))
        }
        return result
    }

    override fun getDefault(): Foc? {
        val external = context.getExternalFilesDir(null)
        return if (external is File) {
            FocAndroid.factory(context, external.absolutePath)
        } else {
            null
        }
    }

    override fun createSolidDirectory(): SolidMapsForgeDirectoryHint {
        val foc: FocFactory = FocAndroidFactory(context)
        val storage: StorageInterface = Storage(context)
        return SolidMapsForgeDirectoryHint(storage, foc, this)
    }

    override fun createSolidFile(): SolidMapsForgeMapFile {
        val factory: FocFactory = FocAndroidFactory(context)
        return SolidMapsForgeMapFile(createSolidDirectory(), factory)
    }

    override fun createSolidRenderTheme(): SolidRenderTheme {
        return SolidRenderTheme(createSolidDirectory(), FocAndroidFactory(context))
    }

    companion object {
        const val MAPS_DIR = "maps"
        const val ORUX_MAPS_DIR = "oruxmaps/mapfiles"
    }
}
